File: LeetCode/Easy/0144_Binary_Tree_Preorder_Traversal/Solution.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// 2021-11-18 (Recursive)
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> path = new ArrayList<Integer>();
        traversePreorder(root, path);
        return path;
    }
    
    private void traversePreorder(TreeNode root, List<Integer> path) {
        if (root == null) return;
        path.add(root.val);
        traversePreorder(root.left, path);
        traversePreorder(root.right, path);
    }
}

// 2021-11-18 (Iterative)
class Solution2 {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> path = new ArrayList<Integer>();
        Deque<TreeNode> stack = new ArrayDeque<TreeNode>();
        
        if (root != null) {
            stack.addFirst(root);
        }
        
        while (!stack.isEmpty()) {
            TreeNode current = stack.pop();
            path.add(current.val);
            if (current.right != null) {
                stack.addFirst(current.right);
            }
            if (current.left != null) {
                stack.addFirst(current.left);
            }
        }
        
        return path;
    }
}

File: LeetCode/Easy/0144_Binary_Tree_Preorder_Traversal/Solution.cpp
// 2022-01-08 (iterative DFS)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        /*
            input: a binary tree
            output: the pre-order traversal of the tree

            notes:
            - preorder is the node, then children

            approach:
            - use a stack to process left, then right subtrees
                - add right child, then left so that left gets processed first
            - store all traversed nodes into a vector and return
        */
        if (!root)
            return {};
            
        vector<int> output;
        deque<TreeNode*> stack = {root};
        while (!stack.empty()) {
            auto node = stack.back();
            stack.pop_back();
            output.push_back(node->val);

            if (node->right)
                stack.push_back(node->right);
            if (node->left)
                stack.push_back(node->left);
        }
        return output;
    }
};